### PR TITLE
Add draft for lora text encoder scale

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Draft PR to show how we could correctly deal with LoRA scale for the text encoder

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -294,6 +294,7 @@ def _encode_prompt(
         negative_prompt=None,
         prompt_embeds: Optional[torch.FloatTensor] = None,
         negative_prompt_embeds: Optional[torch.FloatTensor] = None,
+        lora_scale: Optional[float] = None,
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:297]: Wouldn't be more prudent to pass it from the `load_lora_weights()`. 

IIUC, `_encode_prompt()` is called internally in the pipeline and the end user doesn't have any access to it. So, I don't think this way we expose an explicit control to user to specify what scale they wanna use for the text encoder LoRA.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -319,6 +320,10 @@ def _encode_prompt(
                 weighting. If not provided, negative_prompt_embeds will be generated from `negative_prompt` input
                 argument.
         """
+        # set lora scale so that text encoder can correctly access it
+        if lora_scale is not None and isinstance(self, LoraLoaderMixin):
+            self._lora_scale = lora_scale
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:325]: You meant this to be `_scale`, I guess.

File: src/diffusers/loaders.py [hunk 3]
```diff
@@ -839,6 +839,9 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
         weight_name = kwargs.pop("weight_name", None)
         use_safetensors = kwargs.pop("use_safetensors", None)
 
+        # set lora scale to a reasonable default
+        self._scale = 1
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:843]: But we'd want the users to also specify this explicitly if they want to, no? 

For that, don't you think exposing an argument for it makes sense?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 1 above)
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:297]: I am struggling to understand why do we need to pass this to `_encode_prmompt()`. 

LoRA scale for the text encoder is applied after the LoRA weights for the text encoder is loaded.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -319,6 +320,10 @@ def _encode_prompt(
                 weighting. If not provided, negative_prompt_embeds will be generated from `negative_prompt` input
                 argument.
         """
+        # set lora scale so that text encoder can correctly access it
+        if lora_scale is not None and isinstance(self, LoraLoaderMixin):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:324]: Also, per the docstrings, `lora_scale` is a float no?

File: src/diffusers/loaders.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/loaders.py:843]: Yes, actually users should never set them argument themselves (which is why it's marked private with a `_`). This should just be used to make it work with LoRA

Comment (patrickvonplaten) [reply src/diffusers/loaders.py:843]: I think we have a misunderstanding here, let's maybe take this offline in a quick call next week :-)

**Comment by patrickvonplaten** [comment]
@sayakpaul the purpose of this PR is to create a mechanism that allows the user to change the lora scale of the text encoder. I've adapted the PR to make it functional and used it with your example of a1111 here:

```py
#!/usr/bin/env python3
from diffusers import StableDiffusionPipeline, KDPM2DiscreteScheduler, StableDiffusionImg2ImgPipeline, HeunDiscreteScheduler, KDPM2AncestralDiscreteScheduler, DDIMScheduler,  DPMSolverMultistepScheduler
import time
import os
import torch

path = "gsdf/Counterfeit-V2.5"

pipe = StableDiffusionPipeline.from_pretrained(path, torch_dtype=torch.float16)
pipe.scheduler = DPMSolverMultistepScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)
pipe = pipe.to("cuda")

pipe.load_lora_weights(".", weight_name="light_and_shadow.safetensors")

prompt = "masterpiece, best quality, 1girl, at dusk"
negative_prompt = ("(low quality, worst quality:1.4), (bad anatomy), (inaccurate limb:1.2), "
                   "bad composition, inaccurate eyes, extra digit, fewer digits, (extra arms:1.2), large breasts")

pipe.enable_xformers_memory_efficient_attention()
images = pipe(prompt=prompt, 
    negative_prompt=negative_prompt, 
    width=512, 
    height=768, 
    num_inference_steps=15, 
    num_images_per_prompt=4,
    cross_attention_kwargs={"scale": 0.5},
    generator=torch.manual_seed(0)
).images
```

Note how `cross_attention_kwargs={"scale": 0.5}` now changes both the scale of the text encoder and the unet. Does this PR make more sense now?

**Comment by sayakpaul** [comment]
> Note how cross_attention_kwargs={"scale": 0.5} now changes both the scale of the text encoder and the unet.

But now, we are assuming users will always want to use the same `scale` values for the UNet and the text encoder. It actually might not be the case, no?

File: docs/source/en/training/lora.mdx [hunk 5]
```diff
@@ -260,6 +260,14 @@ pipe.load_lora_weights(lora_model_id)
 image = pipe("A picture of a sks dog in a bucket", num_inference_steps=25).images[0]
 ```
 
+<Tip>
+
+If your LoRA parameters involve the UNet as well as the Text Encoder, then passing
+`cross_attention_kwargs={"scale": 0.5}` will apply the `scale` value to both the UNet 
+and the Text Encoder.
```
Comment (sayakpaul) [inline docs/source/en/training/lora.mdx:267]: Documentation.

File: examples/dreambooth/train_dreambooth_lora.py [hunk 6]
```diff
@@ -870,6 +870,9 @@ def main(args):
         temp_pipeline = DiffusionPipeline.from_pretrained(
             args.pretrained_model_name_or_path, text_encoder=text_encoder
         )
+        # Setting the `_lora_scale` explicitly because we are not using
+        # `load_lora_weights()`.
+        temp_pipeline._lora_scale = 1.0
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:875]: Otherwise, forward pass with the text encoder won't be possible.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:297]: @patrickvonplaten if we do this then we cannot have `# copied from ...` statements in the other pipelines as you can see on the CI.

File: src/diffusers/loaders.py [hunk 7]
```diff
@@ -953,6 +956,10 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
             warn_message = "You have saved the LoRA weights using the old format. To convert the old LoRA weights to the new format, you can first load them in a dictionary and then create a new dictionary like the following: `new_state_dict = {f'unet'.{module_name}: params for module_name, params in old_state_dict.items()}`."
             warnings.warn(warn_message)
 
+    @property
+    def lora_scale(self):
+        return self._lora_scale
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:961]:
```suggestion
        return self._lora_scale if hasattr(self, "_lora_scale") else 1
```

File: examples/dreambooth/train_dreambooth_lora.py (hunk 6 above)
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora.py:875]:
```suggestion
```

File: docs/source/en/training/lora.mdx (hunk 5 above)
Comment (patrickvonplaten) [reply docs/source/en/training/lora.mdx:267]: Nice!

File: src/diffusers/loaders.py [hunk 8]
```diff
@@ -953,6 +956,12 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
             warn_message = "You have saved the LoRA weights using the old format. To convert the old LoRA weights to the new format, you can first load them in a dictionary and then create a new dictionary like the following: `new_state_dict = {f'unet'.{module_name}: params for module_name, params in old_state_dict.items()}`."
             warnings.warn(warn_message)
 
+    @property
+    def lora_scale(self):
+        # property function that returns the lora scale which can be set at run time by the pipeline.
+        # if _lora_scale has not been set, return 1
+        return self._lora_scale if hasattr(self, "_lora_scale") else 1
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:963]:
```suggestion
        return self._lora_scale if hasattr(self, "_lora_scale") else 1.0
```

File: tests/models/test_lora_layers.py [hunk 9]
```diff
@@ -291,6 +292,10 @@ def test_text_encoder_lora_monkey_patch(self):
         del text_attn_procs
         gc.collect()
 
+        # set `_lora_scale` explicitly as we're not directly calling the
+        # pipeline here.
+        pipe._lora_scale = 1.0
+
```
Comment (patrickvonplaten) [inline tests/models/test_lora_layers.py:298]:
```suggestion
```

File: tests/models/test_lora_layers.py [hunk 10]
```diff
@@ -339,6 +344,10 @@ def test_text_encoder_lora_remove_monkey_patch(self):
...
```
Comment (patrickvonplaten) [inline tests/models/test_lora_layers.py:350]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
@sayakpaul, I've now added the LoRALoader to all models where IMO it makes sense. For some pipelines like pix2pix_zero it didn't make much sense so I just added an import so that the copied from still works. IMO it's more important that code stays in sync than to have 100% clean code for this edge case. 

Will merge now so that we have results on the slow tests tomorrow
